graph/simple/weighted_directed.go

```diff
@@ -0,0 +1,285 @@
+// NewWeightedDirectedGraph returns a WeightedDirectedGraph with the specified self and absent
+// edge weight values.
+func NewWeightedDirectedGraph(self, absent float64) *WeightedDirectedGraph {
+	return &WeightedDirectedGraph{
+		nodes: make(map[int64]graph.Node),
+		from:  make(map[int64]map[int64]graph.WeightedEdge),
+		to:    make(map[int64]map[int64]graph.WeightedEdge),
+
+		self:   self,
+		absent: absent,
+
+		nodeIDs: newIDSet(),
+	}
+}
+
+// NewNode returns a new unique Node to be added to g. The Node's ID does
+// not become valid in g until the Node is added to g.
+func (g *WeightedDirectedGraph) NewNode() graph.Node {
+	if len(g.nodes) == 0 {
+		return Node(0)
+	}
+	if int64(len(g.nodes)) == maxInt {
+		panic("simple: cannot allocate node: no slot")
+	}
+	return Node(g.nodeIDs.newID())
+}
+
+// AddNode adds n to the graph. It panics if the added node ID matches an existing node ID.
+func (g *WeightedDirectedGraph) AddNode(n graph.Node) {
+	if _, exists := g.nodes[n.ID()]; exists {
+		panic(fmt.Sprintf("simple: node ID collision: %d", n.ID()))
+	}
+	g.nodes[n.ID()] = n
+	g.from[n.ID()] = make(map[int64]graph.WeightedEdge)
+	g.to[n.ID()] = make(map[int64]graph.WeightedEdge)
+	g.nodeIDs.use(n.ID())
+}
+
+// RemoveNode removes n from the graph, as well as any edges attached to it. If the node
+// is not in the graph it is a no-op.
+func (g *WeightedDirectedGraph) RemoveNode(n graph.Node) {
+	if _, ok := g.nodes[n.ID()]; !ok {
+		return
+	}
+	delete(g.nodes, n.ID())
+
+	for from := range g.from[n.ID()] {
+		delete(g.to[from], n.ID())
+	}
+	delete(g.from, n.ID())
+
+	for to := range g.to[n.ID()] {
+		delete(g.from[to], n.ID())
+	}
+	delete(g.to, n.ID())
+
+	g.nodeIDs.release(n.ID())
+}
+
+// NewWeightedEdge returns a new Edge from the source to the destination node.
+func (g *WeightedDirectedGraph) NewWeightedEdge(from, to graph.Node, weight float64) graph.WeightedEdge {
+	return &WeightedEdge{F: from, T: to, W: weight}
+}
+
+// SetWeightedEdge adds e, an edge from one node to another. If the nodes do not exist, they are added.
+// It will panic if the IDs of the e.From and e.To are equal.
+func (g *WeightedDirectedGraph) SetWeightedEdge(e graph.WeightedEdge) {
+	var (
+		from = e.From()
+		fid  = from.ID()
+		to   = e.To()
+		tid  = to.ID()
+	)
+
+	if fid == tid {
+		panic("simple: adding self edge")
+	}
+
+	if !g.Has(from) {
+		g.AddNode(from)
+	}
+	if !g.Has(to) {
+		g.AddNode(to)
+	}
+
+	g.from[fid][tid] = e
+	g.to[tid][fid] = e
+}
+
+// RemoveEdge removes e from the graph, leaving the terminal nodes. If the edge does not exist
+// it is a no-op.
+func (g *WeightedDirectedGraph) RemoveEdge(e graph.Edge) {
+	from, to := e.From(), e.To()
+	if _, ok := g.nodes[from.ID()]; !ok {
+		return
+	}
+	if _, ok := g.nodes[to.ID()]; !ok {
+		return
+	}
+
+	delete(g.from[from.ID()], to.ID())
+	delete(g.to[to.ID()], from.ID())
+}
+
+// Node returns the node in the graph with the given ID.
+func (g *WeightedDirectedGraph) Node(id int64) graph.Node {
```

Comment:
This method would go a long way to relieve the issues described in gonum/graph#87. Perhaps we should consider adding a `Node(id int64) graph.Node` method to the other graph.Graph implementations?

Then edge creation would be `g.NewEdge(g.Node(xid), g.Node(yid))` rather than `g.NewEdge(simple.Node(xid), simple.Node(yid))`.

But these changes can and should be part of a dedicated PR, to keep this one focused on one thing.